Fix `function-url-no-scheme-relative` false negatives for `@import`

pamelalozano16 opened the pull request:
Closes #8574
> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

Mouvedia commented on .changeset/curly-boats-drop.md:
```diff
@@ -0,0 +1,5 @@
+"stylelint": minor
+---
+
+Fix function-url-no-scheme-relative false negatives for @import
```
```suggestion
Fixed: `function-url-no-scheme-relative` false negatives for `@import`
```

Mouvedia commented on lib/rules/function-url-no-scheme-relative/README.md:
````diff
@@ -9,6 +9,13 @@ a { background-image: url('//www.google.com/file.jpg'); }
  *  This scheme-relative url */
 ```
 
+<!-- prettier-ignore -->
+```css
+@import url('//www.google.com/file.jpg');
+/**         ↑
+ *  This scheme-relative url */
+```
+
````
not necessary
i.e. one example is the norm

ybiquitous replied:
Agree, please remove this example.

Mouvedia commented on lib/rules/function-url-no-scheme-relative/index.mjs:
```diff
@@ -41,6 +41,27 @@ const rule = (primary) => {
+		root.walkAtRules('import', (atRule) => {
+			const params = atRule.params;
+
+			functionArgumentsSearch(params.toLowerCase(), 'url', (args, index) => {
+				const url = args.trim().replace(/^['"]+|['"]+$/g, '');
+
+				if (!isStandardSyntaxUrl(url) || !url.startsWith('//')) {
+					return;
+				}
+
+				report({
+					message: messages.rejected,
+					node: atRule,
+					index,
+					endIndex: index + args.length,
+					result,
+					ruleName,
+				});
+			});
```
refactor
minor

check l27–42
i.e. you could reuse the same function
e.g.
```js
const foo = (node) =>  (args, index) => {
	const url = args.trim().replace(/^['"]+|['"]+$/g, '');

	if (!isStandardSyntaxUrl(url) || !url.startsWith('//')) {
		return;
	}

	report({
		message: messages.rejected,
		node,
		index,
		endIndex: index + args.length,
		result,
		ruleName,
	});
}
```

romainmenke replied:
[Note] Such functions are often named `check` in other rules

ybiquitous reviewed (commented):
@pamelalozano16 Thanks for addressing the reviews. The last one thing, the reported range for `@import url(...)` seems incorrect. Can you take a look? We can see it on [this online demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBLAtgBwgJwC4AIBXbAGwAoByAeioHd6A6AcwgiZJgcjSoDMUODAFYYmFAJQBuADoA7OQEN8wOfnwAjBWADWTbBEKyo8IqTLSQNerWat2nbnwGcRTC1LkBfEABoQ-DgA5BTQ4RBgADxCMQTAAZzjfcAhZfiYEEBVZNQtsQg44ixMstRz-QzBcFBSAWmISGtkIGriwAAsYUJrsGBIFKoA3GCL8XDzh7PxPLyTIVJQmADEcNH6MoTiUpNgMRMQS-As43ABPDhIUWVwRizbcXF34GgxdBgxsBlkYWipjs97LrgkABGABsAGYAJywdQWHyqQ4gP7nQE1OZpFq4BRGBTYKA3EB9XAwY4WGaeIA).

<img width="476" alt="image" src="https://github.com/user-attachments/assets/60cd83df-5611-4592-98d1-ab36ee0cdbbf" />

jeddy3 commented on lib/rules/function-url-no-scheme-relative/README.md:
````diff
@@ -32,6 +32,11 @@ a {
 
+<!-- prettier-ignore -->
+```css
+@import url("//www.google.com/file.jpg");
````
```suggestion
@import url("//www.example.org/foo.css");
```

jeddy3 commented on lib/rules/function-url-no-scheme-relative/__tests__/index.mjs:
```diff
@@ -69,6 +69,21 @@ testRule({
 			description: 'ignore variable',
 		},
+		{
+			code: "@import url('../file.jpg');",
```
```suggestion
			code: "@import url('../foo.css');",
```

jeddy3 commented on lib/rules/function-url-no-scheme-relative/__tests__/index.mjs:
```diff
@@ -69,6 +69,21 @@ testRule({
+			code: "@import url('../file.jpg');",
+		},
+		{
+			code: '@import url("../file.jpg");',
```
```suggestion
			code: '@import url("../foo.css");',
```

jeddy3 commented on lib/rules/function-url-no-scheme-relative/__tests__/index.mjs:
```diff
@@ -69,6 +69,21 @@ testRule({
+			code: "@import url('/path/to/file.jpg');",
+		},
+		{
+			code: "@import url('https://www.google.com/file.jpg');",
```
```suggestion
			code: "@import url('https://www.example.org/foo.css');",
```

jeddy3 commented on lib/rules/function-url-no-scheme-relative/__tests__/index.mjs:
```diff
@@ -69,6 +69,21 @@ testRule({
+			code: "@import url('https://www.google.com/file.jpg');",
+		},
+		{
+			code: "@import url('http://www.google.com/file.jpg');",
```
```suggestion
			code: "@import url('http://www.example.org/foo.css');",
```

jeddy3 commented on lib/rules/function-url-no-scheme-relative/__tests__/index.mjs:
```diff
@@ -128,6 +143,22 @@ testRule({
 			endColumn: 69,
 		},
+		{
+			code: "@import url('//www.google.com/file.jpg')",
```
```suggestion
			code: "@import url('//www.example.org/foo.css')",
```

jeddy3 reviewed (commented):
@pamelalozano16 Thanks for the pull request.

I've made some commitable suggestions so that docs and tests [follow our conventions](https://stylelint.io/developer-guide/rules#write-the-readme). 

From [the spec](https://drafts.csswg.org/css-cascade-5/#at-import):

> The `@import` rule allows users to import style rules from other style sheets.

So, we'll want the `@import`s to import CSS files rather than images.

pamelalozano16 commented:
> @pamelalozano16 Thanks for addressing the reviews. The last one thing, the reported range for `@import url(...)` seems incorrect. Can you take a look? We can see it on [this online demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBLAtgBwgJwC4AIBXbAGwAoByAeioHd6A6AcwgiZJgcjSoDMUODAFYYmFAJQBuADoA7OQEN8wOfnwAjBWADWTbBEKyo8IqTLSQNerWat2nbnwGcRTC1LkBfEABoQ-DgA5BTQ4RBgADxCMQTAAZzjfcAhZfiYEEBVZNQtsQg44ixMstRz-QzBcFBSAWmISGtkIGriwAAsYUJrsGBIFKoA3GCL8XDzh7PxPLyTIVJQmADEcNH6MoTiUpNgMRMQS-As43ABPDhIUWVwRizbcXF34GgxdBgxsBlkYWipjs97LrgkABGABsAGYAJywdQWHyqQ4gP7nQE1OZpFq4BRGBTYKA3EB9XAwY4WGaeIA).
> 
> <img alt="image" width="476" src="https://private-user-images.githubusercontent.com/473530/445368193-60cd83df-5611-4592-98d1-ab36ee0cdbbf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc5MzU4NjEsIm5iZiI6MTc0NzkzNTU2MSwicGF0aCI6Ii80NzM1MzAvNDQ1MzY4MTkzLTYwY2Q4M2RmLTU2MTEtNDU5Mi05OGQxLWFiMzZlZTBjZGJiZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyMlQxNzM5MjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNzllNzdlNWVjZDYzZjNmZDE2Mzk0MmU1NzhlNjI4NmI1NTRhM2RmNzhjNWE3NjhmMjYwZDM2ZmM0NzkxMjQxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.387Yo5_uiH3DOOs41DjRBC9zaDjcHIw_U8_D0m6Q5ek">

It looks like there's a bug in `functionArgumentsSearch` — it doesn’t account for the at-rules' length without parameters, and returns the start and end index of `node.params` instead of the full node. I’ve added a patch to this rule for fixing this behavior. Let me know if that's ok

ybiquitous commented:
@pamelalozano16 Still, the reported range for `@import` is inaccurate, see [this demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBLAtgBwgJwC4AIBXbAGwAoByAeioHd6A6AcwgiZJgcjSoDMUODAFYYmFAJQBuADoA7EABoQ-DgDkAhmjiIYAD00ZBYAM7HF4CLP5MEIYHPz5pIbIQ7Hn8fPdmPHz3kJZMFwUSwBaYhJw2Qhw4zAACxgtcOwYEnVQgDcYT3xcVzzffABfOVLzSCsUJgAxHDQs2yFjS3NYDDNEH38QY1wATw4SFFlcfOdE3Fwu+BoMAGsmBgxsBlkYWioB4Yyx3CR1AE4AdgBGU7AAI2cFByd+oZGD8OrreNx1WSh1bChJiBMrgYANnBUQKUgA).

<img width="324" alt="image" src="https://github.com/user-attachments/assets/4dcd6ee5-b518-4f2f-8b04-07277f1ff385" />

I wonder if we could resolve the issue by fixing the `atRule` string like this, but what do you think?

```diff
 		root.walkAtRules('import', (atRule) => {
-			reportSchemeRelativeUrl(atRule, atRule.params, result);
+			reportSchemeRelativeUrl(atRule, atRule.toString(), result);
 		});
```

pamelalozano16 commented:
> @pamelalozano16 Still, the reported range for `@import` is inaccurate, see [this demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBLAtgBwgJwC4AIBXbAGwAoByAeioHd6A6AcwgiZJgcjSoDMUODAFYYmFAJQBuADoA7EABoQ-DgDkAhmjiIYAD00ZBYAM7HF4CLP5MEIYHPz5pIbIQ7Hn8fPdmPHz3kJZMFwUSwBaYhJw2Qhw4zAACxgtcOwYEnVQgDcYT3xcVzzffABfOVLzSCsUJgAxHDQs2yFjS3NYDDNEH38QY1wATw4SFFlcfOdE3Fwu+BoMAGsmBgxsBlkYWioB4Yyx3CR1AE4AdgBGU7AAI2cFByd+oZGD8OrreNx1WSh1bChJiBMrgYANnBUQKUgA).
> 
> I wonder if we could resolve the issue by fixing the `atRule` string like this, but what do you think?
> 
> ```diff
>  		root.walkAtRules('import', (atRule) => {
> -			reportSchemeRelativeUrl(atRule, atRule.params, result);
> +			reportSchemeRelativeUrl(atRule, atRule.toString(), result);
>  		});
> ```

I believe that works!  (See [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIACBLAtgBwgJwC4AIBXbAGwAoByAeioHd6A6AcwgiZJgcjSoDMUODAFYYmFAJQBuADoA7EABoQ-DgDkAhmjiIYAD00ZBYAM7HF4CLP5MEIYHPz5pIbIQ7Hn8fPdmPHz3kJZMFwUSwBaYhJw2Qhw4zAACxgtcOwYEnVQgDcYT3xcVzzffABfOVLzSCsUJgAxHDQs2yFjS3NYDDNEH38QY1wATw4SFFlcfOdE3Fwu+BoMAGsmBgxsBlkYWioB4Yyx3CQAZl4ATgAmdRgABmcFByd+oZGD8OrreNx1WSh1bChJiBMrgYANnBUQKUgA)). 

<img width="447" alt="Screenshot 2025-05-27 at 1 19 33 PM" src="https://github.com/user-attachments/assets/8ea51dcb-b959-4857-bc2a-1a49e642e685" />

jeddy3 commented on lib/rules/function-url-no-scheme-relative/README.md:
````diff
@@ -32,6 +32,11 @@ a {
 
+<!-- prettier-ignore -->
+```css
+@import url("//www.google.com/foo.scss");
````
```suggestion
@import url("//www.google.com/foo.css");
```

jeddy3 commented on lib/rules/function-url-no-scheme-relative/README.md:
````diff
@@ -54,3 +59,13 @@ a {
+
+<!-- prettier-ignore -->
+```css
+@import url("/path/to/foo.scss");
````
```suggestion
@import url("/path/to/foo.css");
```

jeddy3 commented on lib/rules/function-url-no-scheme-relative/__tests__/index.mjs:
```diff
@@ -69,6 +69,21 @@ testRule({
 			description: 'ignore variable',
 		},
+		{
+			code: "@import url('../foo.scss');",
+		},
+		{
+			code: '@import url("../foo.scss");',
+		},
+		{
+			code: "@import url('/path/to/foo.scss');",
+		},
+		{
+			code: "@import url('https://www.google.com/foo.scss');",
+		},
+		{
+			code: "@import url('http://www.google.com/foo.scss');",
```
```suggestion
			code: "@import url('../foo.css');",
		},
		{
			code: '@import url("../foo.css");',
		},
		{
			code: "@import url('/path/to/foo.css');",
		},
		{
			code: "@import url('https://www.google.com/foo.css');",
		},
		{
			code: "@import url('http://www.google.com/foo.css');",
```

jeddy3 commented on lib/rules/function-url-no-scheme-relative/__tests__/index.mjs:
```diff
@@ -128,6 +143,22 @@ testRule({
 			endColumn: 69,
 		},
+		{
+			code: "@import url('//www.google.com/foo.scss')",
+			message: messages.rejected,
+			line: 1,
+			column: 13,
+			endLine: 1,
+			endColumn: 40,
+		},
+		{
+			code: `@import url("//www.google.com/foo.scss") screen and (min-width: 768px);`,
+			message: messages.rejected,
+			line: 1,
+			column: 13,
+			endLine: 1,
+			endColumn: 40,
+		},
```
```suggestion
			code: "@import url('//www.google.com/foo.css')",
			message: messages.rejected,
			line: 1,
			column: 13,
			endLine: 1,
			endColumn: 39,
		},
		{
			code: `@import url("//www.google.com/foo.css") screen and (min-width: 768px);`,
			message: messages.rejected,
			line: 1,
			column: 13,
			endLine: 1,
			endColumn: 39,
		},
```
